libgpos/include/gpos/memory/CMemoryPoolAlloc.h

```diff
@@ -1,51 +1,58 @@
 //---------------------------------------------------------------------------
 //	Greenplum Database
-//	Copyright (C) 2009-2010 Greenplum Inc.
-//	Copyright (C) 2011 EMC Corp.
+//	Copyright (c) 2004-2015 Pivotal Software, Inc.
 //
 //	@filename:
-//		CMemoryPoolMalloc.h
+//		CMemoryPoolAlloc.h
 //
 //	@doc:
-//		Implementation of memory pool that allocates memory by calling malloc.
+//		Implementation of memory pool using a configurable allocator
 //
 //	@owner:
 //
 //	@test:
 //
 //---------------------------------------------------------------------------
-#ifndef GPOS_CMemoryPoolMalloc_H
-#define GPOS_CMemoryPoolMalloc_H
+#ifndef GPOS_CMemoryPoolAlloc_H
+#define GPOS_CMemoryPoolAlloc_H
 
 #include "gpos/base.h"
-#include "gpos/memory/CMemoryPool.h"
+#include "gpos/memory/CAutoMemoryPool.h"
 
 namespace gpos
 {
 	// memory pool wrapping clib allocation API
-	class CMemoryPoolMalloc : public CMemoryPool
+	class CMemoryPoolAlloc : public CMemoryPool
 	{
 		private:
 
 			// private copy ctor
-			CMemoryPoolMalloc(CMemoryPoolMalloc &);
+			CMemoryPoolAlloc(CMemoryPoolAlloc &);
+			void* (*m_pfnAlloc) (SIZE_T);
+			void (*m_pfnFree) (void*);
 
 		public:
 
 			// ctor
-			CMemoryPoolMalloc()
+			CMemoryPoolAlloc
+				(
+					void* (*pfnAlloc)(SIZE_T),
+					void (*pfnFree)(void*)
+				)
 				:
 				CMemoryPool
 					(
 					NULL /*pmpUnderlying*/,
 					false /*fOwnsUnderlying*/,
 					true /*fThreadSafe*/
```

Comment:
This is no longer thread safe if we decide to use palloc. Even for GPDB vmem system, we have to be very careful as it will not be thread safe.